### PR TITLE
DUPLO-32863: Enhance ecs update_image to target specific container

### PR DESCRIPTION
**Pull request by shantanuduplo**
## Describe Changes

Enhanced the command to allow specifying which container's image to update by name, enabling targeted updates for multi-container task definitions.

## Link to Issues  

https://app.clickup.com/t/8655600/DUPLO-32863

## PR Review Checklist  

- [x] Thoroughly reviewed on local machine.
- [x] Have you added any tests
- [x] Make sure to note changes in Changelog

___

### **PR Type**
Enhancement

___

### **Description**
- Enhanced ECS `update_image` command to target specific containers

- Added optional `container` parameter for multi-container task definitions

- Added comprehensive test coverage for new functionality

- Updated documentation and changelog

___

### **Changes diagram**

```mermaid
flowchart LR
  A["ECS Service"] --> B["Task Definition"]
  B --> C["Container Selection"]
  C --> D["Specific Container"]
  C --> E["First Container (default)"]
  D --> F["Update Image"]
  E --> F
```

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>ecs_service.py</strong><dd><code>Enhanced update_image method for container targeting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/duplo_resource/ecs_service.py

<li>Added optional <code>container</code> parameter to <code>update_image</code> method<br> <li> Implemented logic to find and update specific container by name<br> <li> Enhanced documentation with usage examples and parameter descriptions<br> <li> Improved error handling and return type documentation

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/157/files#diff-00061cbcd7c679c36b38e0b31a9dbe5c3d1f9be94c11602332a5330f96126458">+19/-8</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_ecs_service.py</strong><dd><code>Added comprehensive test coverage for update_image</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/tests/test_ecs_service.py

<li>Added comprehensive test suite for <code>update_image</code> functionality<br> <li> Created tests for specific container targeting scenarios<br> <li> Added tests for default behavior (first container update)<br> <li> Included edge case testing for services without configurations

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/157/files#diff-a4cbb173d0196c88567a454c0cb9f6641ebe4d4cd2ffdaedf92d80e0081dbccf">+95/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Updated changelog with enhancement details</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

CHANGELOG.md

<li>Added entry documenting the ECS update_image enhancement<br> <li> Categorized change under "Added" section for version tracking

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/157/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-32863 Enhance ecs update_image to support updating a specific container in a task definition](https://app.clickup.com/t/8655600/DUPLO-32863)

File: src/duplo_resource/ecs_service.py [hunk 1]
```diff
@@ -158,26 +158,37 @@ def update_taskdef(self,
   @Command()
   def update_image(self, 
                    name: args.NAME, 
+                   container: args.CONTAINER,
```
Comment (kferrone) [inline src/duplo_resource/ecs_service.py:161]: Make this come after the image. Image is the main arg, container is secondary.

File: src/duplo_resource/ecs_service.py [hunk 2]
```diff
@@ -158,26 +158,37 @@ def update_taskdef(self,
...
-    """Update the image for an ECS service.
+    """Update the image for an ECS service container.
 
-    Example:
-      CLI usage
+    Usage: Basic CLI Use
       ```sh
-      duploctl ecs update_image my-service my-image
+        duploctl ecs update_image <service-name> <service-image>
+        duploctl ecs update_image <service-name> --container <container-name> <service-image>
```
Comment (kferrone) [inline src/duplo_resource/ecs_service.py:168]: This is a second example and should be in it's own code block. Better for copy paste this way.

File: src/duplo_resource/ecs_service.py [hunk 3]
```diff
@@ -158,26 +158,37 @@ def update_taskdef(self,
...
-        DuploError: If the ECS service could not be updated.
+      DuploError: If the ECS service could not be updated.
     """
     name = self.prefixed_name(name)
     tdf = self.find_def(name) 
-    tdf["ContainerDefinitions"][0]["Image"] = image
+    if container:
+      for i, container_def in enumerate(tdf["ContainerDefinitions"]):
+        if container_def["Name"] == container:
+          tdf["ContainerDefinitions"][i]["Image"] = image
```
Comment (kferrone) [inline src/duplo_resource/ecs_service.py:188]: Here is a simplified version of the same code: 
```python
for container_def in tdf["ContainerDefinitions"]:
    if container_def["Name"] == container:
        container_def["Image"] = image
```
When using direct loops Python will properly bind objects. No need to set values by referencing the index.

File: src/duplo_resource/ecs_service.py [hunk 4]
```diff
@@ -158,26 +158,37 @@ def update_taskdef(self,
...
+          break
+    else:
+      tdf["ContainerDefinitions"][0]["Image"] = image
```
Comment (kferrone) [inline src/duplo_resource/ecs_service.py:191]: This else statement would prevent us from ever using the `image` argument from this method. 
So this is now impossible: 
```
duploctl ecs update_image myservice myimage:latest --container foo otherimage:latest
```
You can simply always check if the `image` arg is set, and if so do this
```python
tdf["ContainerDefinitions"][0]["Image"] = image
```

If someone were to be weird and do this: 
```
duploctl ecs update_image myservice nginx:latest --container myservice somethingelse:latest
```
Just let it get set twice and say whatever. Another options is you could be nice and throw a warning that the `myservice` container is getting set twice. I just don't think anyone will do something like this.
